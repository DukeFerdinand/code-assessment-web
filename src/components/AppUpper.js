import React from 'react'
import PropTypes from 'prop-types'
import ReactSVG from  'react-svg'
import { connect } from 'react-redux'

import { getTotal, getItemCount } from '../reducers'
import { toggleModal } from '../actions'
import CartIcon from '../assets/cart.svg'

export const AppUpper = ({ title, toggleModal, total, itemCount }) => {
  const linkText = (total, itemCount) => {
    if (itemCount) {
      if (total.toString().indexOf('.') === -1) {
        total = `${total.toString()}.00`
      }
      // Borrowed this layout from NewEgg
      // Looks like [icon] $9.00 (1 items)
      return `$${total} (${itemCount} item${itemCount > 1 ? 's' : ''})`
    }
    return 'Your cart is empty'
  }
  return (
    <div className="app-title-wrapper">
      <h2 className="app-title">{title}</h2>
      <span
        className="cart-link"
        onClick={() => toggleModal()}
      >
        <ReactSVG
          wrapper="span"
          className="cart-icon"
          src={CartIcon}
        />
        <p className="cart-link-text" >{linkText(total, itemCount)}</p>
      </span>
    </div>
  )
}

AppUpper.PropTypes = {
  title: PropTypes.string.isRequired,
  toggleModal: PropTypes.func
}

const mapStateToProps = (state) => {
  return {
    total: getTotal(state),
    itemCount: getItemCount(state)
  }
}

export default connect(mapStateToProps, { toggleModal })(AppUpper)