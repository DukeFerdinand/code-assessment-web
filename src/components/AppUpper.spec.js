import React from 'react'
import { shallow } from 'enzyme'
// importing non-redux-connected variant
import { AppUpper } from './AppUpper';

const setup = (title, total = 0.00, itemCount = 0) => {
  const actions = {
    toggleModal: jest.fn()
  }
  const component = shallow(
    <AppUpper title={title} total={total} itemCount={itemCount} {...actions} />
  )
  return {
    component,
    actions,
    cartLink: component.find('p.cart-link-text')
  }
}

describe('AppUpper component', () => {
  describe('when given title prop', () => {
    it('should display the app title', () => {
      const { component } = setup('App Title');
      expect(component.find('h2.app-title').text()).toBe('App Title');
    })
  });
  describe('when given a total and item count', () => {
    it('should render cart link with single item', () => {
      const { cartLink } = setup('App Title', 10.01, 1)
      expect(cartLink.text()).toBe('$10.01 (1 item)')
    })

    it('should render cart link with multiple items', () => {
      const { cartLink } = setup('App Title', 10.01, 3)
      expect(cartLink.text()).toBe('$10.01 (3 items)')
    })

    describe('when given a total with no decimals', () => {
      it('should render two decimals', () => {
        const { cartLink } = setup('App Title', 9, 1)
        expect(cartLink.text()).toBe('$9.00 (1 item)')
      })
    })
  });
  describe('when not given total or item count', () => {
    it('should render empty cart message', () => {
      const { cartLink } = setup('App Title')
      expect(cartLink.text()).toBe('Your cart is empty')
    })
  });
  it('should match the snapshot', () => {
    const { component } = setup('App Title')
    expect(component).toMatchSnapshot();
  })
});