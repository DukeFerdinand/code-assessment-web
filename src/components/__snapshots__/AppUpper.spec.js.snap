// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppUpper component should match the snapshot 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="app-title-wrapper"
  >
    <h2
      className="app-title"
    >
      App Title
    </h2>
    <span
      className="cart-link"
      onClick={[Function]}
    >
      <ReactSVG
        className="cart-icon"
        evalScripts="never"
        fallback={null}
        loading={null}
        onInjected={[Function]}
        renumerateIRIElements={true}
        src="cart.svg"
        svgClassName={null}
        svgStyle={Object {}}
        wrapper="span"
      />
      <p
        className="cart-link-text"
      >
        Your cart is empty
      </p>
    </span>
  </div>,
  "nodes": Array [
    <div
      className="app-title-wrapper"
    >
      <h2
        className="app-title"
      >
        App Title
      </h2>
      <span
        className="cart-link"
        onClick={[Function]}
      >
        <ReactSVG
          className="cart-icon"
          evalScripts="never"
          fallback={null}
          loading={null}
          onInjected={[Function]}
          renumerateIRIElements={true}
          src="cart.svg"
          svgClassName={null}
          svgStyle={Object {}}
          wrapper="span"
        />
        <p
          className="cart-link-text"
        >
          Your cart is empty
        </p>
      </span>
    </div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {},
      "_currentElement": <AppUpper
        itemCount={0}
        title="App Title"
        toggleModal={[MockFunction]}
        total={0}
      />,
      "_debugID": 11,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "itemCount": 0,
          "title": "App Title",
          "toggleModal": [MockFunction],
          "total": 0,
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 6,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div
          className="app-title-wrapper"
        >
          <h2
            className="app-title"
          >
            App Title
          </h2>
          <span
            className="cart-link"
            onClick={[Function]}
          >
            <ReactSVG
              className="cart-icon"
              evalScripts="never"
              fallback={null}
              loading={null}
              onInjected={[Function]}
              renumerateIRIElements={true}
              src="cart.svg"
              svgClassName={null}
              svgStyle={Object {}}
              wrapper="span"
            />
            <p
              className="cart-link-text"
            >
              Your cart is empty
            </p>
          </span>
        </div>,
        "_debugID": 12,
        "_renderedOutput": <div
          className="app-title-wrapper"
        >
          <h2
            className="app-title"
          >
            App Title
          </h2>
          <span
            className="cart-link"
            onClick={[Function]}
          >
            <ReactSVG
              className="cart-icon"
              evalScripts="never"
              fallback={null}
              loading={null}
              onInjected={[Function]}
              renumerateIRIElements={true}
              src="cart.svg"
              svgClassName={null}
              svgStyle={Object {}}
              wrapper="span"
            />
            <p
              className="cart-link-text"
            >
              Your cart is empty
            </p>
          </span>
        </div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <AppUpper
    itemCount={0}
    title="App Title"
    toggleModal={[MockFunction]}
    total={0}
  />,
}
`;
